client-go watch: context support

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The Lister and Watcher interfaces only supported methods without context, but
were typically implemented with client-go API calls which need a context. New
interfaces get added using the same approach as in
https://github.com/kubernetes/kubernetes/pull/129109.

For StreamWatcher and RaceFreeFakeWatcher it was easier to add `WithLogger`
configuration methods which override the default global logger. The downside of
this approach is that it cannot be enforced by logcheck. This is okay for
Kubernetes in this case because all existing usages of those two types are
local to client-go and get updated immediately.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/pull/129125

#### Special notes for your reviewer:

Generated informers: for compatibility reasons, the old functions without the ctx parameter still
get generated, now with context.Background instead of context.TODO. In practice
that code won't be used by the client-go reflector code because it prefers
the *WithContext functions, but it cannot be ruled out that some other code
only supports the old fields.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

/wg structured-logging

pohly committed:
client-go informers: provide ListWatch *WithContext variants

For compatibility reasons, the old functions without the ctx parameter still
get generated, now with context.Background instead of context.TODO. In practice
that code won't be used by the client-go reflector code because it prefers
the *WithContext functions, but it cannot be ruled out that some other code
only supports the old fields.

pohly committed:
client-go + apimachinery watch: context support

The Lister and Watcher interfaces only supported methods without context, but
were typically implemented with client-go API calls which need a context. New
interfaces get added using the same approach as in
https://github.com/kubernetes/kubernetes/pull/129109.

pohly commented:
/test pull-kubernetes-unit 

Unrelated error, should be fixed by https://github.com/kubernetes/kubernetes/pull/129615.

pohly commented:
/test pull-kubernetes-apidiff

pohly commented:
/test pull-kubernetes-e2e-kind-ipv6 

Diverse timeouts.

sttts commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -27,50 +27,156 @@ import (
+	WatchWithContext(ctx context.Context, options metav1.ListOptions) (watch.Interface, error)
+}
+
+func Watcher2Context(w Watcher) WatcherWithContext {
```
as for lister

sttts commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -27,50 +27,156 @@ import (
 type ListFunc func(options metav1.ListOptions) (runtime.Object, error)
 
+// ListFuncWithContext knows how to list resources
+type ListFuncWithContext func(ctx context.Context, options metav1.ListOptions) (runtime.Object, error)
```
nit: ListWithContextFunc ?

pohly replied:
Changed.

sttts commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -27,50 +27,156 @@ import (
 	WatchFunc WatchFunc
+
+	ListFuncWithContext  ListFuncWithContext
+	WatchFuncWithContext WatchFuncWithContext
```
doc the semantics if both are set or just one variant

sttts replied:
Related below I guess: `// For backward compatibility, all function fields are populated.`

pohly replied:
Updated the comment with a different wording.

sttts commented on staging/src/k8s.io/client-go/tools/watch/informerwatcher.go:
```diff
@@ -137,13 +152,31 @@ func NewIndexerInformerWatcher(lw cache.ListerWatcher, objType runtime.Object) (
+			case <-ctx.Done():
+			case <-w.StopChan():
+			}
+			w.Stop()
```
why do we need that after `<-w.StopChan()`?

pohly replied:
That was the "Optionally map cancellation..." comment above. It's only relevant if the context is done. While it doesn't hurt to always call it, it may be more obvious why it gets called when its in the `case <-ctx.Done()` - moved and updated the comment.

pohly replied:
Removed entirely as part of the NewIndexerInformerWatcherWithContext -> NewIndexerInformerWatcherWithLogger simplification (to be confirmed, see https://github.com/kubernetes/kubernetes/pull/129341#discussion_r1935556454).

sttts commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -317,15 +318,7 @@ func (rw *RetryWatcher) ResultChan() <-chan watch.Event {
-	default:
-		close(rw.stopChan)
-	}
+	rw.cancel(errors.New("asked to stop"))
```
nice

pohly commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -27,50 +27,156 @@ import (
+	ListWithContext(ctx context.Context, options metav1.ListOptions) (runtime.Object, error)
+}
+
+func Lister2Context(l Lister) ListerWithContext {
```
Changed. I'm updating https://github.com/kubernetes/kubernetes/pull/129109 accordingly - it uses the same pattern.

soltysh commented on staging/src/k8s.io/apimachinery/pkg/watch/streamwatcher.go:
```diff
@@ -59,8 +60,16 @@ type StreamWatcher struct {
+}
+
+// NewStreamWatcherWithLogger creates a StreamWatcher from the given decoder and logger.
+func NewStreamWatcherWithLogger(logger klog.Logger, d Decoder, r Reporter) *StreamWatcher {
```
I'd like to avoid a situation where we have `*WithLogger` and then `*WithContext` like we have in https://github.com/kubernetes/kubernetes/blob/07cc2308c679f5402931ee46879a9edee36f8a4c/staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go#L46-L93 although that is a bit different situation. Anyway, `StreamWatcher` has a `done` channel, which we allow users to invoke through `Stop` method. So the question is do we want to just pass a context and wire it such that both `Stop` and `ctx.Done` are equivalent? We've done something similar in https://github.com/kubernetes/kubernetes/blob/07cc2308c679f5402931ee46879a9edee36f8a4c/staging/src/k8s.io/client-go/tools/cache/controller.go#L143-L148 for example.

pohly replied:
A `NewStreamWatcherWithContext` could use a similar construct as `NewIndexerInformerWatcherWithContext` in this PR to map context cancellation to `Stop`. It definitely would be more consistent and IMHO useful.

I don't remember any particular reason for deciding not to do that also here. I'll change it.

pohly replied:
However, such a change then begs the question: should the context that is getting passed to this `NewStreamWatcherWithContext` be the same context that is passed to `Request.Watch`?

https://github.com/kubernetes/kubernetes/blob/07cc2308c679f5402931ee46879a9edee36f8a4c/staging/src/k8s.io/client-go/rest/request.go#L756-L761

The call chain is `Watch -> watchInternal -> newStreamWatcher -> NewStreamWatcherWithLogger`.

That would be a functional change of `Watch` with regards to how it uses the context that it is being passed. Someone might currently pass a short-lived context that gets canceled after `Watch` returns, which then breaks the returned StreamWatcher.

So `newStreamWatcher` would have to call `NewStreamWatcherWithContext` with `klog.NewContext(context.Background(), klog.FromContext(ctx))`. Doable, but more complicated.

This problem does not exist for `NewIndexerInformerWatcherWithContext` because where the PR passes the parent's context the watcher is expected to stop before the caller returns:
https://github.com/kubernetes/kubernetes/blob/07cc2308c679f5402931ee46879a9edee36f8a4c/staging/src/k8s.io/client-go/tools/watch/until.go#L128-L135

Should `NewStreamWatcherWithContext` perhaps get replaced by a simpler `NewStreamWatcherWithLogger`?

soltysh replied:
Watch is and always will be long running function, even the doc says on Watch method says:
```
// Watch attempts to begin watching the requested location.
```
so it's a begin, not a short-lived request. So I'm inclined to say it's not a functional change, but maybe that's a question for sig-apimachinery?

pohly replied:
It doesn't say anything about the context, though. Usually, contexts are only passed through functions call, so my expectation as a Go developer is that the context is only used while the `Watch` call runs.

pohly replied:
For example, code might set a timeout like this to stop the attempt to set up the watch:
```
ctx, cancel := context.Timeout(time.Minute)
defer cancel()
return req.Watch(ctx)
```

pohly replied:
> Should NewIndexerInformerWatcherWithContext perhaps get replaced by a simpler NewIndexerInformerWatcherWithLogger?

I convinced myself that this is the right way to go and therefore I included it in this PR as an additional commit. I can squash or remove it, depending on how we decide to do this.

pohly replied:
Can you LGTM again? My last commit merely updated the unit test for the no longer supported context cancellation instead of removing it completely - [fixed](https://github.com/kubernetes/kubernetes/compare/91f8403520a4a54b22993d08a93e5cd58484bf97..1a8d8c9b4a33daf9330434e1ad544ef3571722a3). Sorry, should have checked test results before pinging you!

soltysh commented on staging/src/k8s.io/apimachinery/pkg/watch/watch.go:
```diff
@@ -103,29 +104,42 @@ func (w emptyWatch) ResultChan() <-chan Event {
+func NewFakeWithOptions(options FakeOptions) *FakeWatcher {
 	return &FakeWatcher{
-		result: make(chan Event, size),
+		logger: ptr.Deref(options.Logger, klog.Background()),
```
Identical comment as the other wrt passing a context rather than logger here, which you can then wire to stopping the watcher.

pohly committed:
client-go watch: NewIndexerInformerWatcherWithContext -> WithLogger

The ability to automatically stop on context cancellation was new functionality
that adds complexity and wasn't really used in Kubernetes. If someone wants
this, they can add it outside of the function.

A *WithLogger variant avoids the complexity and is consistent with
NewStreamWatcherWithLogger over in apimachinery.

soltysh reviewed (approved):
/lgtm

pohly commented:
/assign @deads2k @sttts 

For approval.

sttts commented:
/lgtm
/approve
